src/coreclr/vm/precode_portable.hpp

```diff
@@ -0,0 +1,122 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+
+#ifndef __PRECODE_PORTABLE_H__
+#define __PRECODE_PORTABLE_H__
+
+#ifndef FEATURE_PORTABLE_ENTRYPOINTS
+#error Requires FEATURE_PORTABLE_ENTRYPOINTS to be set
+#endif // !FEATURE_PORTABLE_ENTRYPOINTS
+
+class PortableEntryPoint final
+{
+public: // static
+    static bool IsNativeEntryPoint(TADDR addr);
+    static MethodDesc* GetMethodDesc(TADDR addr);
+    static void* GetInterpreterData(TADDR addr);
+    static void SetInterpreterData(TADDR addr, PCODE interpreterData);
+
+private: // static
+    static PortableEntryPoint* ToPortableEntryPoint(TADDR addr);
+
+private:
+    INDEBUG(size_t _canary);
+    void* _pActualCode;
```

Comment:
I think we want `_pActualCode` to be first (before _canary) since the offset is going to be part of the ABI for AOT compiled code.